File: lib/api/services.ts
import { apiClient } from "../api-client"

export interface Service {
  id: string
  name: string
  description: string
  duration: string
  price: number
  is_active: boolean
  created_at: string
  updated_at: string
}

export interface ServiceListResponse {
  total: number
  items: Service[]
}

export interface ServiceCreate {
  name: string
  description: string
  duration: string
  price: number
  is_active?: boolean
}

export interface ServiceUpdate {
  name?: string
  description?: string
  duration?: string
  price?: number
  is_active?: boolean
}

export const servicesApi = {
  /**
   * Get all services
   */
  getAll: async (params?: {
    skip?: number
    limit?: number
    is_active?: boolean
  }): Promise<ServiceListResponse> => {
    const searchParams = new URLSearchParams()
    if (params?.skip !== undefined) searchParams.append("skip", params.skip.toString())
    if (params?.limit !== undefined) searchParams.append("limit", params.limit.toString())
    if (params?.is_active !== undefined) searchParams.append("is_active", params.is_active.toString())

    return apiClient<ServiceListResponse>(`/services?${searchParams.toString()}`)
  },

  /**
   * Get service by ID
   */
  getById: async (id: string): Promise<Service> => {
    return apiClient<Service>(`/services/${id}`)
  },

  /**
   * Create service (admin only)
   */
  create: async (data: ServiceCreate): Promise<Service> => {
    return apiClient<Service>("/services", {
      method: "POST",
      body: JSON.stringify(data),
    })
  },

  /**
   * Update service (admin only)
   */
  update: async (id: string, data: ServiceUpdate): Promise<Service> => {
    return apiClient<Service>(`/services/${id}`, {
      method: "PATCH",
      body: JSON.stringify(data),
    })
  },

  /**
   * Delete service (admin only)
   */
  delete: async (id: string): Promise<void> => {
    return apiClient(`/services/${id}`, {
      method: "DELETE",
    })
  },
}

File: ruqya-backend/app/api/v1/services.py
from fastapi import APIRouter, Depends, HTTPException, status, Query
from sqlalchemy.orm import Session
from typing import Optional
import uuid

from app.database import get_db
from app.models.service import Service
from app.models.user import User
from app.schemas.service import (
    ServiceCreate,
    ServiceUpdate,
    ServiceResponse,
    ServiceListResponse,
)
from app.core.security import get_current_user, get_current_admin_user

router = APIRouter(prefix="/services", tags=["Services"])


@router.get("", response_model=ServiceListResponse)
async def get_services(
    skip: int = Query(0, ge=0),
    limit: int = Query(20, ge=1, le=100),
    is_active: Optional[bool] = Query(None),
    db: Session = Depends(get_db)
):
    """Get all services with optional filtering."""
    query = db.query(Service)
    
    if is_active is not None:
        query = query.filter(Service.is_active == is_active)
    
    total = query.count()
    services = query.offset(skip).limit(limit).all()
    
    return ServiceListResponse(
        total=total,
        items=[ServiceResponse.model_validate(s) for s in services]
    )


@router.post("", response_model=ServiceResponse, status_code=status.HTTP_201_CREATED)
async def create_service(
    service_data: ServiceCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_admin_user)
):
    """Create a new service (Admin only)."""
    new_service = Service(
        id=str(uuid.uuid4()),
        **service_data.model_dump()
    )
    
    db.add(new_service)
    db.commit()
    db.refresh(new_service)
    
    return ServiceResponse.model_validate(new_service)


@router.get("/{service_id}", response_model=ServiceResponse)
async def get_service(service_id: str, db: Session = Depends(get_db)):
    """Get a specific service by ID."""
    service = db.query(Service).filter(Service.id == service_id).first()
    
    if not service:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Service not found"
        )
    
    return ServiceResponse.model_validate(service)


@router.patch("/{service_id}", response_model=ServiceResponse)
async def update_service(
    service_id: str,
    service_data: ServiceUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_admin_user)
):
    """Update a service (Admin only)."""
    service = db.query(Service).filter(Service.id == service_id).first()
    
    if not service:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Service not found"
        )
    
    # Update fields
    update_data = service_data.model_dump(exclude_unset=True)
    for field, value in update_data.items():
        setattr(service, field, value)
    
    db.commit()
    db.refresh(service)
    
    return ServiceResponse.model_validate(service)


@router.delete("/{service_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_service(
    service_id: str,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_admin_user)
):
    """Delete a service (Admin only)."""
    service = db.query(Service).filter(Service.id == service_id).first()
    
    if not service:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Service not found"
        )
    
    db.delete(service)
    db.commit()
    
    return None
    